platform/engineering/design-docs/micro-frontend-architecture.md

```diff
@@ -0,0 +1,255 @@
+# Micro Front end Architecture
+
+**Author(s):** Chris Valarida (chris.valarida@adhocteam.us)  
+**Last Updated:** February XX, 2020  
+**Status:** **Draft** | In Review | Approved  
+**Approvers:** Andrew Gunsch, Rian Fowler, Wyatt Walter
+
+<!-- Should the approvers include anybody from DEPO? -->
+
+
+## Overview
+
+### Objective
+To establish a micro front end architecture which enables application teams to
+operate independently.
+
+This document is intended for front end and DevOps engineers on the
+Veteran-facing Services Platform (VSP) and Veteran-Facing Services (VSA) teams.
+
+
+### Background
+
+<!-- Could put a definitions section here if it becomes useful -->
+
+#### The State of `vets-website`
+
+##### Build
+The `vets-website` build is responsible for a lot of things. The most important
+pieces include:
+- Pulling content from the [CMS](#CMS)
+- Building that content into HTML pages
+- Validating the content for broken links, accessibility violations, etc.
+- Building all JavaScript bundles
+  - Both the common bundle and individual application bundles
+- Building all CSS bundles
+  - Both common and application-specific
+- Running all >5k unit tests for all applications
+- Running all end-to-end tests for all applications
+
+
+##### Deploy
+When `vets-website` is deployed, all content and applications are deployed
+together. **There is no way to deploy one application without deploying
+everything.**
+
+There is, however, a way to deploy the latest content from the CMS without
+deploying the latest application code. This is called a **content-only deploy**.
+The latest `vets-website` release is used instead of building code in the latest
+commit of the `vets-website` `master` branch.
+
+
+#### The Problems
+##### Failures in one Service affect all other Services
+When one application team introduces test failures into `master`, all PRs from
+all teams that include the offending commit in `master` will fail in Continuous
+Integration (CI) because all tests are run, not just the tests for a single
+application.
+
+**Example:** Team A introduces a test failure in `master`. Team B has an urgent
```

Comment:
how often does this happen? doesn't CI generally prevent failing tests from getting checked in?

edit: I see your comment below about two conflicting changes. unless that's happening a lot, that seems like a pretty rare case to be the justification for this big of an overhaul

Reply:
In the last few months, this has happened maybe half a dozen times at least. It's not super rare.

Reply:
Cases worth adding (note to self):
- When a team writes a test that will inherently change over time
  - Such as the one [this PR](https://github.com/department-of-veterans-affairs/vets-website/pull/11793) addressed
- When a team writes a test that modifies a global property without resetting it
  - The failure looks something like [this](http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-website/detail/master/6136/tests)
  - It may not be obvious, and it may only break tests given a certain `CHOMA_SEED`